File: pylang/LinkedList/LinkedList.py
class Node:
    def __init__(self, NewNode):
        self.data = NewNode
        self.NextNode = None

class LinkedList:
    def __init__(self):
        self.Head = None

    def AppendNode(self, Data):
        # global Head
        NewNode = Node(Data)

        if self.Head == None:
            self.Head = NewNode
        else:
            Tail = self.Head
            while Tail.NextNode:
                Tail = Tail.NextNode
            
            Tail.NextNode = NewNode
    
    def InsertAfter(self, Current, Data):
        NewNode = Node(Data)
        NewNode.NextNode = Current.NextNode
        Current.NextNode = NewNode

    def InsertBefore(self, Current, Data):
        # global Head
        NewNode = Node(Data)

        if self.Head == None:
            self.Head = NewNode
        elif self.Head == Current:
            NewNode.NextNode = self.Head
            self.Head = NewNode
        else:
            Prev = self.Head
            while Prev is not None and Prev.NextNode != Current:
                Prev = Prev.NextNode
            if Prev is not None:
                Prev.NextNode = NewNode
                NewNode.NextNode = Current

    def InsertNewHead(self, Data):
        # global Head
        NewHead = Node(Data)

        if self.Head == None:
            self.Head = NewHead
        else:
            NewHead.NextNode = self.Head
            self.Head = NewHead

    def RemoveNode(self, Remove):
        # global Head

        if self.Head is Remove:
            self.Head = self.Head.NextNode
        else:
            Current = self.Head
            while Current is not None and Current.NextNode is not Remove:
                Current = Current.NextNode
            if Current is not None:
                Current.NextNode = Remove.NextNode

    def GetNodeAt(self, Location):
        # global Head
        Location -= 1
        Current = self.Head

        while Current is not None and Location >= 0:
            Current = Current.NextNode
            Location -= 1
        
        return Current

    def PrintList(self):
        cur = self.Head
        while cur:
            print(cur.data)
            cur = cur.NextNode

# Head = None  # 전역 변수로 선언

print("\n연결리스트 테스트\n")
ll = LinkedList()
ll.AppendNode(10)
ll.AppendNode(30)
ll.AppendNode(20)
ll.AppendNode(40)
ll.AppendNode(50)

# 출력
# cur = Head
# while cur:
#     print(cur.data)
#     cur = cur.NextNode
ll.PrintList()

print("\n데이터 삽입 후\n")
Current = ll.GetNodeAt(2)
ll.InsertAfter(Current, 3000)
ll.InsertBefore(Current, 5000)

# cur = Head
# while cur:
#     print(cur.data)
#     cur = cur.NextNode
ll.PrintList()

print("\nHead 변경\n")
ll.InsertNewHead(5)
# cur = Head
# while cur:
#     print(cur.data)
#     cur = cur.NextNode
ll.PrintList()

print("\n노드 삭제\n")
remove = ll.GetNodeAt(5)
ll.RemoveNode(remove)
# cur = Head
# while cur:
#     print(cur.data)
#     cur = cur.NextNode
ll.PrintList()

File: clang/LinkedList/LinkedList.c
#include "LinkedList.h"

// 노드 생성
Node* SLL_CreateNode(ElementType NewData)
{
    Node* NewNode = (Node*)malloc(sizeof(Node));

    NewNode->Data = NewData; // 데이터를 저장
    NewNode->NextNode = NULL; // 다음 노드에 대한 포인터는 NULL로 초기화

    return NewNode; // 노드의 주소를 반환
}

// 노드 소멸
void SLL_DestroyNode(Node* Node)
{
    free(Node);
}

void SLL_DestroyAllNodes(Node** List)
{
    Node* Current = *List;
    Node* Next;

    while ( Current != NULL )
    {
        Next = Current->NextNode;
        free(Current);
        Current = Next;
    }

    *List = NULL;
}

// 노드 추가
void SLL_AppendNode(Node** Head, Node* NewNode)
{
    // 헤드 노드가 NULL이라면 새로운 노드가 Head가 됨
    if ( (*Head) == NULL )
    {
        *Head = NewNode;
    }
    else
    {
        // 테일을 찾아 NewNode를 연결
        Node* Tail = (*Head);
        while ( Tail->NextNode != NULL )
        {
            Tail = Tail->NextNode;
        }

        Tail->NextNode = NewNode;
    }
}

// 노드 삽입
void SLL_InsertAfter(Node* Current, Node* NewNode)
{
    NewNode->NextNode = Current->NextNode;
    Current->NextNode = NewNode;
}

// 추가
void SLL_InsertBefore(Node** Head, Node* Current, Node* NewHead)
{
    if ( (*Head) == NULL )
    {
        (*Head) = NewHead;
    }
    else if ( *Head == Current )
    {
        NewHead->NextNode = *Head;
        *Head = NewHead;
    }
    else
    {
        Node* Prev = *Head;
        while ( Prev != NULL && Prev->NextNode != Current)
        {
            Prev = Prev->NextNode;
        }

        if ( Prev != NULL )
        {
            Prev->NextNode = NewHead;
            NewHead->NextNode = Current;
        }
    }
}

void SLL_InsertNewHead(Node** Head, Node* NewHead)
{
    if ( (*Head) == NULL )
    {
        (*Head) = NewHead;
    }
    else
    {
        NewHead->NextNode = (*Head);
        (*Head) = NewHead;
    }
}

// 노드 제거
void SLL_RemoveNode(Node** Head, Node* Remove)
{
    if ( (*Head) == Remove )
    {
        *Head = Remove->NextNode;
    }
    else
    {
        Node* Current = *Head;
        while ( Current != NULL && Current->NextNode != Remove )
        {
            Current = Current->NextNode;
        }

        if ( Current != NULL )
            Current->NextNode = Remove->NextNode;
    }
}

// 노드 탐색
Node* SLL_GetNodeAt(Node* Head, int Location)
{
    Node* Current = Head;

    while ( Current != NULL && (--Location) >= 0 )
    {
        Current = Current->NextNode;
    }

    return Current;
}

// 노드 개수 세기
int SLL_GetNodeCount(Node* Head)
{
    int Count = 0;
    Node* Current = Head;

    while ( Current != NULL )
    {
        Current = Current->NextNode;
        Count++;
    }

    return Count;
}